contracts/contracts/UserReplicaSetManager.sol

```diff
@@ -0,0 +1,135 @@
+      require(msg.sender == _proposerWallet, "Invalid sender");
+      // Requirements for non deployer address
+      if (msg.sender != deployer) {
+          require(spIdToCreatorNode[_proposerSpId] == _proposerWallet, "Mismatch proposer wallet for existing spID");
+          require(spIdToCreatorNode[_proposerSpId] != address(0x00), "Unregistered sender spID");
+      }
+
+      // TODO: Event
+      spIdToCreatorNode[_newCnodeId] = _newCnodeDelegateOwnerWallet;
+    }
+
+    // Function used to permission updates to a given user's replica set 
+    function updateReplicaSet(
+        uint _userId,
+        uint _primary,
+        uint[] calldata _secondaries,
+        uint _oldPrimary,
+        uint[] calldata _oldSecondaries) external
+      {
+          // Caller's notion of existing primary must match regisered value on chain
+          require(artistReplicaSets[_userId].primary == _oldPrimary,  "Invalid prior primary configuration"); 
+
+          // Get user object from UserFactory
+          UserFactoryInterface userFactory = UserFactoryInterface(registry.getContract(userFactoryRegistryKey));
+          require(userFactory.userExists(_userId), "Valid user required");
+
+          (address userWallet, ) = userFactory.getUser(_userId);
+
+          // A valid updater can be one of the dataOwnerWallet, existing creator node, or contract deployer
+          bool validUpdater = false;
+          // TODO: Replace msg.sender below with recovered signature from _subjectSig object
+          if (msg.sender == userWallet || msg.sender == spIdToCreatorNode[_oldPrimary] || msg.sender == deployer) {
+              validUpdater = true;
+          }
+          // Caller's notion of secondary values must match registered value on chain
+          // A secondary node can also be considered a valid updater
+          require(_oldSecondaries.length == artistReplicaSets[_userId].secondaries.length, "Invalid prior secondary configuration");
+          for (uint i = 0; i < _oldSecondaries.length; i++) {
+              require(
+                  artistReplicaSets[_userId].secondaries[i] == _oldSecondaries[i],
+                  "Invalid prior secondary configuration"
+              );
+
+              // A valid updater has been found
+              if (msg.sender == spIdToCreatorNode[_oldSecondaries[i]]) {
+                validUpdater = true;
+              }
+          }
+          require(validUpdater == true, "Invalid update operation");
+
+          // Perform replica set update
+          artistReplicaSets[_userId] = ReplicaSet({
+              primary: _primary,
+              secondaries: _secondaries
+          });
+
+          // Confirm primary and every incoming secondary is valid
+          require(spIdToCreatorNode[_primary] != address(0x00), "Primary must exist"); 
+          for (uint i = 0; i < _secondaries.length; i++) {
+              require(spIdToCreatorNode[_secondaries[i]] != address(0x00), "Secondary must exist");
+          }
```

Comment:
does it matter with tx atomicity? whether it happens there or here no gas is expended and state is reverted if a single id maps to an invalid wallet

Reply:
yeah it won't affect the outcome bc of atomic tx, just logically easier to follow in order. technically there is a gas difference since only unused gas is sent back in revert, but thats minimal + doesn't rly matter